File: android/core/network/src/main/kotlin/com/rendezvousil/core/network/RendezvousRepository.kt
package com.rendezvousil.core.network

import com.rendezvousil.core.network.dto.RatesPayload
import com.rendezvousil.core.network.dto.WeatherPayload
import com.rendezvousil.core.schedule.ScheduleNowNext
import com.rendezvousil.core.schedule.model.Announcement
import com.rendezvousil.core.schedule.model.Meal
import com.rendezvousil.core.schedule.model.NowNextResult
import com.rendezvousil.core.schedule.model.SchedulePayload
import com.rendezvousil.core.schedule.model.VolunteerScheduleSlot
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class RendezvousRepository(
    private val apiClient: ApiClient,
    private val fallbackLoader: ScheduleFallbackLoader? = null,
    private val onScheduleSynced: ((SchedulePayload?) -> Unit)? = null,
) {
    private val _schedule = MutableStateFlow<SchedulePayload?>(null)
    val schedule: StateFlow<SchedulePayload?> = _schedule.asStateFlow()

    private val _meals = MutableStateFlow<List<Meal>>(emptyList())
    val meals: StateFlow<List<Meal>> = _meals.asStateFlow()

    private val _volunteerSchedules = MutableStateFlow<Map<String, VolunteerScheduleSlot>>(emptyMap())
    val volunteerSchedules: StateFlow<Map<String, VolunteerScheduleSlot>> =
        _volunteerSchedules.asStateFlow()

    private val _liveAnnouncements = MutableStateFlow<List<Announcement>>(emptyList())
    val liveAnnouncements: StateFlow<List<Announcement>> = _liveAnnouncements.asStateFlow()

    private val _scheduleAnnouncements = MutableStateFlow<List<Announcement>>(emptyList())
    val scheduleAnnouncements: StateFlow<List<Announcement>> = _scheduleAnnouncements.asStateFlow()

    private val _weather = MutableStateFlow<WeatherPayload?>(null)
    val weather: StateFlow<WeatherPayload?> = _weather.asStateFlow()

    private val _rates = MutableStateFlow<RatesPayload?>(null)
    val rates: StateFlow<RatesPayload?> = _rates.asStateFlow()

    private val _isLoadingSchedule = MutableStateFlow(false)
    val isLoadingSchedule: StateFlow<Boolean> = _isLoadingSchedule.asStateFlow()

    private val _scheduleError = MutableStateFlow<String?>(null)
    val scheduleError: StateFlow<String?> = _scheduleError.asStateFlow()

    suspend fun loadScheduleBundle() {
        _isLoadingSchedule.value = true
        _scheduleError.value = null
        try {
            _schedule.value = apiClient.getSchedule()
            syncSharedSnapshot()
        } catch (error: Exception) {
            val bundled = fallbackLoader?.loadBundledSchedule()
            if (bundled != null) {
                _schedule.value = bundled
                _scheduleError.value = null
                syncSharedSnapshot()
            } else {
                _scheduleError.value = error.message
            }
        } finally {
            _isLoadingSchedule.value = false
        }
    }

    suspend fun loadScheduleExtras() = coroutineScope {
        val mealsTask = async { fetchMeals() }
        val volunteerTask = async { fetchVolunteers() }
        val scheduleAnnouncementsTask = async { fetchScheduleAnnouncements() }
        mealsTask.await()
        volunteerTask.await()
        scheduleAnnouncementsTask.await()
    }

    suspend fun loadUpdates() = coroutineScope {
        val announcementsTask = async { fetchLiveAnnouncements() }
        val weatherTask = async { fetchWeather() }
        announcementsTask.await()
        weatherTask.await()
        if (_schedule.value == null) {
            loadScheduleBundle()
        } else {
            syncSharedSnapshot()
        }
    }

    suspend fun loadRates() {
        _rates.value = try {
            apiClient.getRates(year = AppConfig.EVENT_YEAR)
        } catch (_: Exception) {
            null
        }
    }

    fun meal(date: String, mealType: String): Meal? =
        _meals.value.firstOrNull { it.date == date && it.meal_type == mealType }

    fun volunteers(date: String, timeSlot: String): VolunteerScheduleSlot? =
        _volunteerSchedules.value["$date|$timeSlot"]

    fun nowNext(): NowNextResult {
        val items = _schedule.value?.luItems ?: return NowNextResult()
        return ScheduleNowNext.evaluate(items)
    }

    private fun syncSharedSnapshot() {
        onScheduleSynced?.invoke(_schedule.value)
    }

    private suspend fun fetchMeals() {
        _meals.value = try {
            apiClient.getMeals().meals.orEmpty()
        } catch (_: Exception) {
            emptyList()
        }
    }

    private suspend fun fetchVolunteers() {
        _volunteerSchedules.value = try {
            apiClient.getVolunteerSchedule().schedules.orEmpty()
        } catch (_: Exception) {
            emptyMap()
        }
    }

    private suspend fun fetchLiveAnnouncements() {
        _liveAnnouncements.value = try {
            apiClient.getAnnouncements().announcements.orEmpty()
        } catch (_: Exception) {
            emptyList()
        }
    }

    private suspend fun fetchScheduleAnnouncements() {
        _scheduleAnnouncements.value = try {
            apiClient.getScheduleAnnouncements().announcements.orEmpty()
        } catch (_: Exception) {
            emptyList()
        }
    }

    private suspend fun fetchWeather() {
        _weather.value = try {
            apiClient.getWeather()
        } catch (_: Exception) {
            null
        }
    }
}

File: ios/RendezvousIL/Services/RendezvousRepository.swift
import Foundation

@MainActor
@Observable
final class RendezvousRepository {
    var schedule: SchedulePayload?
    var meals: [Meal] = []
    var volunteerSchedules: [String: VolunteerScheduleSlot] = [:]
    var liveAnnouncements: [Announcement] = []
    var scheduleAnnouncements: [Announcement] = []
    var weather: WeatherPayload?
    var rates: RatesPayload?

    var isLoadingSchedule = false
    var isRefreshingSchedule = false
    var isLoadingUpdates = false
    var scheduleError: String?
    var scheduleSource: ScheduleDataSource?
    var lastScheduleRefresh: Date?

    /// True when showing bundled or cached schedule instead of a fresh network copy.
    var isUsingOfflineSchedule: Bool {
        guard schedule != nil else { return false }
        return scheduleSource != .network
    }

    private let weekFrom = "2027-05-03"
    private let weekTo = "2027-05-07"

    init() {
        applyOfflineScheduleIfNeeded()
    }

    /// Offline-first bootstrap — call once after sign-in.
    func bootstrap() async {
        applyOfflineScheduleIfNeeded()
        if AppStoreScreenshotMode.isEnabled {
            // Demo mode: bundled schedule only — no network (faster, deterministic frames).
            return
        }
        async let scheduleTask: Void = loadScheduleBundle()
        async let extrasTask: Void = loadScheduleExtras()
        async let updatesTask: Void = loadUpdates()
        _ = await (scheduleTask, extrasTask, updatesTask)
    }

    func loadScheduleBundle() async {
        applyOfflineScheduleIfNeeded()

        if schedule == nil {
            isLoadingSchedule = true
        } else {
            isRefreshingSchedule = true
        }
        scheduleError = nil
        defer {
            isLoadingSchedule = false
            isRefreshingSchedule = false
        }

        do {
            let payload: SchedulePayload = try await RepositoryFetch.withTimeout {
                try await APIClient.shared.get("/api/schedule")
            }
            schedule = payload
            scheduleSource = .network
            lastScheduleRefresh = Date()
            ScheduleDataStore.saveCached(payload)
            await syncSharedSnapshot()
            try? await ReminderService.shared.rescheduleAll(items: payload.luItems)
        } catch {
            if schedule == nil, let offline = ScheduleDataStore.bestOfflineSchedule() {
                schedule = offline.schedule
                scheduleSource = offline.source
                await syncSharedSnapshot()
            } else if schedule == nil {
                scheduleError = "Could not load schedule. Pull to retry."
            }
            #if DEBUG
            AppLog.bootstrap("schedule fetch failed: \(error.localizedDescription)")
            #endif
        }
    }

    func loadScheduleExtras() async {
        async let mealsTask: Void = fetchMeals()
        async let volunteerTask: Void = fetchVolunteers()
        async let scheduleAnnouncementsTask: Void = fetchScheduleAnnouncements()
        _ = await (mealsTask, volunteerTask, scheduleAnnouncementsTask)
    }

    func loadUpdates() async {
        let showLoading = liveAnnouncements.isEmpty && weather == nil
        if showLoading { isLoadingUpdates = true }
        defer { isLoadingUpdates = false }

        async let announcementsTask: Void = fetchLiveAnnouncements()
        async let weatherTask: Void = fetchWeather()
        _ = await (announcementsTask, weatherTask)
        if schedule == nil {
            await loadScheduleBundle()
        } else {
            await syncSharedSnapshot()
        }
    }

    func loadRates() async {
        do {
            rates = try await RepositoryFetch.withTimeout {
                try await APIClient.shared.get("/api/rates?year=2027")
            }
        } catch {
            rates = nil
        }
    }

    func meal(date: String, mealType: String) -> Meal? {
        meals.first { $0.date == date && $0.meal_type == mealType }
    }

    func volunteers(date: String, timeSlot: String) -> VolunteerScheduleSlot? {
        volunteerSchedules["\(date)|\(timeSlot)"]
    }

    func nowNext() -> NowNextResult {
        guard let items = schedule?.luItems else {
            return NowNextResult(current: nil, next: nil)
        }
        return ScheduleNowNext.evaluate(items: items)
    }

    func syncSharedSnapshot() async {
        let result = nowNext()
        SharedScheduleStore.publish(schedule: schedule, nowNext: result)
        guard NotificationService.shared.liveActivityEnabled else { return }
        Task(priority: .utility) {
            await LiveActivityManager.shared.refresh(schedule: schedule, nowNext: result)
        }
    }

    private func applyOfflineScheduleIfNeeded() {
        guard schedule == nil else { return }
        guard let offline = ScheduleDataStore.bestOfflineSchedule() else { return }
        schedule = offline.schedule
        scheduleSource = offline.source
        SharedScheduleStore.publish(schedule: schedule, nowNext: nowNext())
    }

    private func fetchMeals() async {
        do {
            let response: MealsResponse = try await RepositoryFetch.withTimeout {
                try await APIClient.shared.get("/api/meals")
            }
            meals = response.meals ?? []
        } catch {
            meals = []
        }
    }

    private func fetchVolunteers() async {
        let from = weekFrom
        let to = weekTo
        do {
            let response: VolunteerWeekResponse = try await RepositoryFetch.withTimeout {
                try await APIClient.shared.get(
                    "/api/volunteer-schedule?from=\(from)&to=\(to)"
                )
            }
            volunteerSchedules = response.schedules ?? [:]
        } catch {
            volunteerSchedules = [:]
        }
    }

    private func fetchLiveAnnouncements() async {
        do {
            let response: AnnouncementsResponse = try await RepositoryFetch.withTimeout {
                try await APIClient.shared.get("/api/announcements")
            }
            liveAnnouncements = response.announcements ?? []
        } catch {
            liveAnnouncements = []
        }
    }

    private func fetchScheduleAnnouncements() async {
        do {
            let response: ScheduleAnnouncementsResponse = try await RepositoryFetch.withTimeout {
                try await APIClient.shared.get("/api/announcements/schedule")
            }
            scheduleAnnouncements = response.announcements ?? []
        } catch {
            scheduleAnnouncements = []
        }
    }

    private func fetchWeather() async {
        do {
            weather = try await RepositoryFetch.withTimeout {
                try await APIClient.shared.get("/api/weather")
            }
        } catch {
            weather = nil
        }
    }
}
